1-Draft/RFCxxxx-Secrets-Management.md

```diff
@@ -0,0 +1,288 @@
+
+>[!NOTE]
+>KeyChain support is unlikely to be available in the first release of this feature.
+
+`Default` on Linux is [Gnome Keyring](https://wiki.gnome.org/Projects/GnomeKeyring/).
+
+>[!NOTE]
+>Linux has many options for local credential management.  The choice of Gnome Keyring
+>is that it's a simple local vault we can easily test against and validate that our
+>design works with a well-known Linux local vault.  Since the extension model is
+>open, we expect additional vault support to come from owners of those vaults or
+>the community.
+
+### Credential Vault Extensions
+
+Vault extensions are PowerShell modules that provide implementations of four required functions:
+
+- SetSecret - Adds a secret to the vault
+
+- GetSecret - Retrieves a secret from the vault
+
+- RemoveSecret - Removes a secret from the vault
+
+- GetSecretInfo - Returns information about one or more secrets (but not the secret itself)
+
+The extension module can expose the above functions either through a C# class implementing an abstract type, or by publicly exporting script cmdlet functions.
+Each function implementation takes a set of parameter arguments needed for secret manipulation and error reporting.
+In addition, each function implementation takes an optional parameter argument that is a dictionary of additional parameters, specific to that implementation.
+When registering the vault extension, the additional parameters are stored securely in the built-in local vault as a hashtable.
+At runtime the additional parameters are read from the built-in local vault and passed to the vault implementing functions.
+
+### C# class implementation
+
+The PowerShell module must include a 'RequiredAssemblies' entry in the module manifest which provides the name of the binary that implements the abstract type.
+
+```powershell
+@{
+    ModuleVersion = '1.0'
+    RequiredAssemblies = @('AKVaultExtension')
+}
+```
+
+```C#
+// AKVaultExtension implements these abstract methods
+public abstract bool SetSecret(
+    string name,
+    object secret,
+    IReadOnlyDictionary<string, object> parameters,
+    out Exception error);
+
+public abstract object GetSecret(
+    string name,
+    IReadOnlyDictionary<string, object> parameters,
+    out Exception error);
+
+public abstract bool RemoveSecret(
+    string name,
+    IReadOnlyDictionary<string, object> parameters,
+    out Exception error);
+
+public abstract KeyValuePair<string, string>[] GetSecretInfo(
+    string filter,
+    IReadOnlyDictionary<string, object> parameters,
+    out Exception error);
+```
+
+When PowerShell loads the module, the required assembly will be loaded and the implementing type becomes available.
+
+### PowerShell script implementation
+
+For a script implementation, the PowerShell module must include a subdirectory named 'ImplementingModule' in the same directory containing the module manifest file.

```

Comment:
Any consideration for using classes here instead of lose functions? Class based resources for DSC was a much appreciated improvement.